src/runtime/dsl.ts

```diff
@@ -0,0 +1,431 @@
+//--------------------------------------------------------------------
+// Javascript DSL for writing Eve programs
+//--------------------------------------------------------------------
+
+// There don't seem to be TypeScript definitions for these by default,
+// so here we are.
+declare var Proxy:new (obj:any, proxy:any) => any;
+declare var Symbol:any;
+
+import {RawValue, Register, isRegister, GlobalInterner, Scan, IGNORE_REG, ID,
+        InsertNode, Node, Constraint, FunctionConstraint} from "./runtime";
+import * as runtime from "./runtime";
+import * as indexes from "./indexes";
+
+//--------------------------------------------------------------------
+// Utils
+//--------------------------------------------------------------------
+
+function maybeIntern(value:(RawValue|Register)):Register|ID {
+  if(value === undefined || value === null) throw new Error("Trying to intern an undefined");
+  if(isRegister(value)) return value;
+  return GlobalInterner.intern(value);
+}
+
+//--------------------------------------------------------------------
+// DSLVariable
+//--------------------------------------------------------------------
+
+type DSLValue = RawValue|Register;
+class DSLVariable {
+  static CURRENT_ID = 0;
+  id: number;
+  constructor(public name:string, public value:DSLValue = new Register(-1)) {
+    this.id = DSLVariable.CURRENT_ID++;
+  }
+}
+
+//--------------------------------------------------------------------
+// DSLFunction
+//--------------------------------------------------------------------
+
+class DSLFunction {
+  returnValue:DSLVariable;
+
+  constructor(public block:DSLBlock, public path:string[], public args:any[]) {
+    this.returnValue = new DSLVariable("record");
+    block.registerVariable(this.returnValue);
+  }
+
+  compile() {
+    let constraints:FunctionConstraint[] = [];
+    let result = maybeIntern(this.block.toValue(this.returnValue));
+    let values = this.args.map((v) => maybeIntern(this.block.toValue(v)))
+    constraints.push(FunctionConstraint.create(this.path.join("/"), {result}, values) as FunctionConstraint);
+    return constraints;
+  }
+}
+
+//--------------------------------------------------------------------
+// DSLRecord
+//--------------------------------------------------------------------
+
+class DSLRecord {
+  record: DSLVariable;
+  output: boolean;
+  fields: any;
+  constructor(public block:DSLBlock, public tag:string[], initial:any) {
```

Comment:
Initial should have a comment describing it, or be more descriptively named (attributes, perhaps?).